### PR TITLE
raftstore-v2: active flush before close

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/14981

What's Changed:

```commit-message
active flush before server stop
```

Actively flush tablet for regions whose flush gap (replay count if start now) exceeds the threshold.
To speed up this process, temporarily increase both raftstore threads and flush threads (hight priority background threads in rocksdb).

The log ouput below shows the PR reduces the replay complete time from a minute to less than 5 seconds, with the cost of flush time around 20 seconds during server stop. 
The reduction of raft log replay time could help raft group to be a health status during rolling restart which makes QPS more stable during rolling restart.

Before this PR:
```
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-0 ...
[2023/06/20 17:18:25.896 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=43.631654296s] [paused_peers=1197] [normal_peers=3235]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-1 ...
[2023/06/20 17:15:31.702 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=68.988228697s] [paused_peers=1605] [normal_peers=4361]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-2 ...
[2023/06/20 17:10:59.763 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=30.560611465s] [paused_peers=1544] [normal_peers=5107]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-3 ...
[2023/06/20 17:08:24.533 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=71.347384062s] [paused_peers=1123] [normal_peers=3001]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-4 ...
[2023/06/20 17:05:24.141 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=81.399500498s] [paused_peers=1318] [normal_peers=4284]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-5 ...
[2023/06/20 16:56:09.713 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=66.50860096s] [paused_peers=1013] [normal_peers=2992]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-6 ...
[2023/06/20 16:52:30.143 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=45.825848807s] [paused_peers=1009] [normal_peers=2676]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-7 ...
[2023/06/20 16:49:14.230 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=57.874569055s] [paused_peers=1133] [normal_peers=2958]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-8 ...
[2023/06/20 16:46:17.039 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=63.464762302s] [paused_peers=990] [normal_peers=2720]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-9 ...
[2023/06/20 16:43:05.475 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=42.021027224s] [paused_peers=991] [normal_peers=3144]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-10 ...
[2023/06/20 16:40:09.985 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=81.145752539s] [paused_peers=912] [normal_peers=2956]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-11 ...
[2023/06/20 16:36:39.782 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=63.537761126s] [paused_peers=881] [normal_peers=2484]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-12 ...
[2023/06/20 16:33:33.809 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=71.877206713s] [paused_peers=968] [normal_peers=3154]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-13 ...
[2023/06/20 16:30:56.931 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=37.705788596s] [paused_peers=879] [normal_peers=2275]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-14 ...
[2023/06/20 16:29:07.444 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=35.907338005s] [paused_peers=1015] [normal_peers=2852]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-15 ...
[2023/06/20 16:25:16.425 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=40.203443677s] [paused_peers=979] [normal_peers=3058]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-16 ...
[2023/06/20 16:22:59.976 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=49.459650587s] [paused_peers=948] [normal_peers=3168]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-17 ...
[2023/06/20 16:20:32.360 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=53.260173769s] [paused_peers=1040] [normal_peers=3469]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-18 ...
[2023/06/20 16:17:33.890 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=113.241509325s] [paused_peers=956] [normal_peers=3560]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-19 ...
[2023/06/20 16:14:17.270 +08:00] [INFO] [mod.rs:115] ["The raft log replay completed"] [elapsed=74.128840751s] [paused_peers=1548] [normal_peers=1871]
```

After this PR:
```
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-0 ...
[2023/06/21 11:09:27.456 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.183420873s] [paused_peers=6] [normal_peers=4438]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-1 ...
[2023/06/21 11:06:36.609 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=7.209233252s] [paused_peers=1] [normal_peers=5928]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-2 ...
[2023/06/21 11:02:05.022 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=4.564232613s] [paused_peers=1] [normal_peers=6647]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-3 ...
[2023/06/21 10:57:18.646 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.487190739s] [paused_peers=35] [normal_peers=4086]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-4 ...
[2023/06/21 10:53:27.669 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=5.617190918s] [paused_peers=32] [normal_peers=5548]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-5 ...
[2023/06/21 10:46:41.712 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=3.788757035s] [paused_peers=69] [normal_peers=3862]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-6 ...
[2023/06/21 10:41:43.107 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.307158097s] [paused_peers=93] [normal_peers=3604]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-7 ...
[2023/06/21 10:35:35.228 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=4.318016518s] [paused_peers=37] [normal_peers=4106]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-8 ...
[2023/06/21 10:31:14.712 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=3.825657632s] [paused_peers=62] [normal_peers=3659]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-9 ...
[2023/06/21 10:26:41.822 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=4.357355312s] [paused_peers=69] [normal_peers=4061]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-10 ...
[2023/06/21 10:23:01.320 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.293527403s] [paused_peers=130] [normal_peers=3713]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-11 ...
[2023/06/21 10:19:15.640 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.135150486s] [paused_peers=71] [normal_peers=3307]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-12 ...
[2023/06/21 10:15:49.555 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.333567499s] [paused_peers=89] [normal_peers=4014]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-13 ...
[2023/06/21 10:10:38.598 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.380354002s] [paused_peers=99] [normal_peers=3037]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-14 ...
[2023/06/21 10:07:33.005 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=2.057677119s] [paused_peers=43] [normal_peers=3847]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-15 ...
[2023/06/21 10:04:23.744 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=3.823112393s] [paused_peers=108] [normal_peers=3905]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-16 ...
[2023/06/21 10:01:07.790 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=1.885230626s] [paused_peers=38] [normal_peers=4087]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-17 ...
[2023/06/21 09:58:35.077 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=4.935349169s] [paused_peers=259] [normal_peers=4208]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-18 ...
[2023/06/21 09:53:59.280 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=3.965871646s] [paused_peers=149] [normal_peers=4356]
Searching for keyword 'The raft log replay completed' in log file on tc-tikv-19 ...
[2023/06/21 09:50:14.353 +08:00] [INFO] [[mod.rs:133](http://mod.rs:133/)] ["The raft log replay completed"] [elapsed=4.997857346s] [paused_peers=353] [normal_peers=3248]
```
 
The time used for flush (around 20 seconds)
```
Searching for keyword in log file on tc-tikv-0 ...
[2023/06/21 11:07:07.349 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 11:07:24.985 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-1 ...
[2023/06/21 11:03:07.360 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 11:03:26.599 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-2 ...
[2023/06/21 10:58:37.426 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:58:51.140 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-3 ...
[2023/06/21 10:55:07.420 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:55:25.744 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-4 ...
[2023/06/21 10:47:37.216 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:47:56.759 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-5 ...
[2023/06/21 10:43:08.121 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:43:29.003 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-6 ...
[2023/06/21 10:36:37.864 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:37:03.204 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-7 ...
[2023/06/21 10:32:07.694 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:32:30.619 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-8 ...
[2023/06/21 10:28:07.267 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:28:26.781 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-9 ...
[2023/06/21 10:24:07.196 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:24:23.615 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-10 ...
[2023/06/21 10:20:37.334 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:20:57.102 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-11 ...
[2023/06/21 10:17:07.889 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:17:24.955 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-12 ...
[2023/06/21 10:13:15.556 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:13:33.487 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-13 ...
[2023/06/21 10:08:37.142 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:08:53.569 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-14 ...
[2023/06/21 10:05:13.124 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:05:38.399 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-15 ...
[2023/06/21 10:02:00.870 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 10:02:17.125 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-16 ...
[2023/06/21 09:59:12.265 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 09:59:32.129 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-17 ...
[2023/06/21 09:55:07.365 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 09:55:33.925 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-18 ...
[2023/06/21 09:51:07.424 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 09:51:28.422 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
Searching for keyword in log file on tc-tikv-19 ...
[2023/06/21 09:47:37.226 +08:00] [INFO] [server2.rs:1285] ["flush-before-close: flush begin"]
[2023/06/21 09:47:59.116 +08:00] [INFO] [server2.rs:1317] ["flush-before-close: flush done"]
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/engine_traits/src/flush.rs [hunk 1]
```diff
@@ -103,12 +103,18 @@ impl SstApplyState {
 #[derive(Debug)]
 pub struct FlushState {
     applied_index: AtomicU64,
+
+    // This is only used for flush before server stop.
+    // It provides a direct path for flush progress by letting raftstore directly know the current
+    // flush progress.
+    flushed_index: [AtomicU64; DATA_CFS_LEN],
```
Comment (tonyxuqqi) [inline components/engine_traits/src/flush.rs:110]: If it's optional, it's better to be Option<[AtomicU64; DATA_CFS_LEN]>

Comment (SpadeA-Tang) [reply components/engine_traits/src/flush.rs:110]: It's not optional.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/operation/ready/apply_trace.rs [hunk 2]
```diff
@@ -544,6 +555,74 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+                    let cf = offset_to_cf(i);
+                    apply_trace.on_flush(cf, flush_index);
+                    max_flush_index = u64::max(max_flush_index, flush_index);
+                }
+
+                apply_trace.maybe_advance_admin_flushed(max_flush_index);
+                let admin_flush = apply_trace.admin.flushed;
+                apply_trace.persisted_applied = admin_flush;
+
+                if self.storage().estimate_replay_count() < flush_threshold {
```
Comment (bufferflies) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:602]: it can break direct if this condition is true.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:602]: Yes

**Review by bufferflies — APPROVED** [review]
rest LGTM

File: components/raftstore-v2/src/operation/ready/apply_trace.rs [hunk 3]
```diff
@@ -544,6 +555,75 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+                    let (_, _, tablet_index) = ctx
+                        .tablet_registry
+                        .parse_tablet_name(Path::new(tablet.path()))
+                        .unwrap();
+                    let mut lb = ctx.engine.log_batch(1);
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: Why is this necessary? Isn't flushed index written in EventListener? @SpadeA-Tang

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: Only data cf will be flushed in EventListener.

Comment (tabokie) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: That is a design flaw. We should try to advance admin flushed first thing after restart, then use the refreshed index to read raft logs.

Comment (tabokie) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: Also this PR didn't remove the old flush after start logic.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: > Also this PR didn't remove the old flush-after-start logic.

Only paused peer will trigger flush and I think it's fine.

Comment (tabokie) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: What do you mean? Every region will be flushed once the replay is completed. The code is at `Peer::try_compelete_recovery`.

Comment (tabokie) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: Fixing them here: https://github.com/tikv/tikv/pull/15023

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs:607]: > What do you mean? Every region will be flushed once the replay is completed. The code is at `Peer::try_compelete_recovery`.

image
`pause_for_replay` means the replay count exceeds 128

**Comment by tonyxuqqi** [comment]
Close #14459
